File: Tools/resultsdbpy/resultsdbpy/view/static/js/configuration.js
// Copyright (C) 2019 Apple Inc. All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions
// are met:
// 1. Redistributions of source code must retain the above copyright
//    notice, this list of conditions and the following disclaimer.
// 2. Redistributions in binary form must reproduce the above copyright
//    notice, this list of conditions and the following disclaimer in the
//    documentation and/or other materials provided with the distribution.
//
// THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
// THE POSSIBILITY OF SUCH DAMAGE.

import {queryToParams} from '/assets/js/common.js';

// These are flipped delibrately, it makes the fromQuery function return configurations in an
// intuitive order.
const REQUIRED_MEMBERS = ['architecture', 'version', 'is_simulator', 'platform'];
const OPTIONAL_MEMBERS = ['flavor', 'style', 'model', 'version_name'];
const FILTERING_MEMBERS = ['sdk'];
const VERSION_OFFSET_CONSTANT = 1000;
const SDK_REGEX = /(\d+)([A-Z])(\d+)(.*)/;

// This class is very similar to the Python Configuration class
class Configuration {
    static members() {
        return [...FILTERING_MEMBERS, ...OPTIONAL_MEMBERS, ...REQUIRED_MEMBERS];
    }
    static fromQuery() {
        const params = queryToParams(document.URL.split('?')[1]);
        let configs = [new Configuration()];

        Configuration.members().forEach(member => {
            const argumentsForMember = params[member];
            if (!argumentsForMember || argumentsForMember.length === 0)
                return;

            let originalConfigs = configs;
            configs = [];
            argumentsForMember.forEach(argument => {
                originalConfigs.forEach(config => {
                    let newConfig = new Configuration(config);
                    newConfig[member] = argument;
                    configs.push(new Configuration(newConfig));
                });
            });
        });

        return configs;
    }
    static versionToInteger(version) {
        if (typeof version === 'number')
            return version;
        let versionList = null;
        if (typeof version === 'string')
            versionList = version.split('.');
        else {
            if (!Array.isArray(version))
                return null;
            versionList = version;
        }
        if (versionList.length < 1 || versionList.length > 3)
            return null;

        let result = 0;
        let index = 0;
        while (index < 3) {
            result *= VERSION_OFFSET_CONSTANT;
            if (index < versionList.length)
                result += parseInt(versionList[index]);
            ++index;
        }
        return result;
    }
    static integerToVersion(version) {
        let result = [0, 0, 0];
        for (let index = 0; index < result.length; ++index) {
            result[result.length - (index + 1)] = version % VERSION_OFFSET_CONSTANT;
            version = Math.floor(version / VERSION_OFFSET_CONSTANT);
        }
        return result.join('.');
    }

    static combine()
    {
        if (!arguments.length)
            return new Configuration();
        const combined = new Configuration(arguments[0]);
        for (let i = 1; i < arguments.length; ++i) {
            Configuration.members().forEach(member => {
                if (arguments[i][member] !== combined[member])
                    combined[member] = null;
            });
        }
        return combined;
    }

    constructor(json = {}) {
        this.platform = json.platform ? json.platform : null;
        this.version = json.version ? Configuration.versionToInteger(json.version) : null;
        this.version_name = json.version_name ? json.version_name : null;
        this.sdk = json.sdk ? json.sdk : null;
        if (typeof json.is_simulator === 'string')
            this.is_simulator = json.is_simulator.toLowerCase() === 'true' ? true : false;
        else if (typeof json.is_simulator === 'boolean')
            this.is_simulator = json.is_simulator;
        else
            this.is_simulator = null;
        this.style = json.style ? json.style : null;
        this.flavor = json.flavor ? json.flavor : null;
        this.model = json.model ? json.model : null;
        this.architecture = json.architecture ? json.architecture : null;

        // Mid-year releases really need to be treated with an entirely different version_name. The only way to reliably
        // identify such a release is with the SDK (version numbers have historically ranged between .2 and .4)
        // While appending E to all realese with an SDK after a mid-year release isn't entirely correct, it's close enough
        // that the user can quickly dicern the differences by inspecting the specific SDK differences
        if (this.sdk && this.version_name && !(json instanceof Configuration)) {
            const match = this.sdk.match(SDK_REGEX);
            const ending = this.version_name.substring(this.version_name.length - 2)
            if (match && ending !== ' E' && match[2].localeCompare('E') >= 0)
                this.version_name = `${this.version_name} E`;
            else if (!match)
                console.error(`'${this.sdk}' does not match the SDK regular expression`);
        }
    }
    toKey() {
        let result = ''
        if (this.platform != null)
            result += ' ' + this.platform
        if (this.version_name != null)
            result += ' ' + this.version_name;
        else if (this.version != null)
            result += ' ' + Configuration.integerToVersion(this.version);
        if (this.sdk != null)
            result += ' (' + this.sdk + ')';
        if (this.is_simulator != null && this.is_simulator)
            result += ' Simulator';
        if (this.style != null)
            result += ' ' + this.style;
        if (this.flavor != null)
            result += ' ' + this.flavor;
        if (this.model != null)
            result += ' on ' + this.model;
        if (this.architecture != null)
            result += ' using ' + this.architecture;

        if (result)
            return result.substr(1);
        return 'All';
    }
    toString() {
        let result = '';

        // Version name implies platform and version
        if (this.version_name != null)
            result += ' ' + this.version_name;
        else {
            if (this.platform != null)
                result += ' ' + this.platform
            if (this.version != null)
                result += ' ' + Configuration.integerToVersion(this.version);
        }
        if (this.is_simulator != null && this.is_simulator)
            result += ' Simulator';
        if (this.flavor != null)
            result += ' ' + this.flavor;
        if (this.style != null) {
            result += ' ' + this.style.split('-').map(word => word.charAt(0).toUpperCase() + word.slice(1)).join('-');
        } if (this.model != null)
            result += ' on ' + this.model;
        if (this.architecture != null)
            result += ' with ' + this.architecture;

        if (this.sdk != null)
            result += ' (' + this.sdk + ')';

        if (result)
            return result.substr(1);
        return 'All';
    }
    compare(configuration) {
        if (configuration === null)
            return 1;

        const lst = [['platform'], ['is_simulator'], ['version_name', 'version'], ['flavor'], ['style'], ['architecture'], ['model']];
        let result = 0;
        lst.forEach(keyFamily => {
            // Grouping keys into families allows keys to be ignored if another is defined.
            // For example, if two configurations declare the same version name, despite having a different
            // version number, they will be treated as identical
            let wasKeyFamilyDefined = false;

            keyFamily.forEach(key => {
                if (result || wasKeyFamilyDefined)
                    return;

                if (this[key] !== null && configuration[key] !== null) {
                    if (typeof this[key] === 'string') {
                        if (key === 'version_name' && this[key].startsWith(configuration[key]))
                            result = 0;
                        else
                            result = this[key].localeCompare(configuration[key]);
                    }
                    else if (typeof this[key] === 'number' || typeof this[key] === 'boolean')
                        result = this[key] - configuration[key];
                    else
                        console.error(typeof this[key] + ' is not a recognized configuration type')
                    wasKeyFamilyDefined = true;
                }
            });
        });
        return result;
    }
    compareSDKs(configuration) {
        if (this.sdk === null || configuration.sdk === null)
            return 0;

        const thisMatch = this.sdk.match(SDK_REGEX);
        const configurationMatch = configuration.sdk.match(SDK_REGEX);

        if (!thisMatch && !configurationMatch) {
            console.error(`'${this.sdk}' and '${configuration.sdk}' do not match the SDK regular expression`);
            return this.sdk.localeCompare(configuration.sdk);
        }
        if (!thisMatch && configurationMatch)
            return -1;
        if (thisMatch && !configurationMatch)
            return 1;

        const majorDiff = parseInt(thisMatch[1]) - parseInt(configurationMatch[1]);
        if (majorDiff)
            return majorDiff;
        const minorDiff = thisMatch[2].localeCompare(configurationMatch[2]);
        if (minorDiff)
            return minorDiff;
        const buildDiff = parseInt(thisMatch[3]) - parseInt(configurationMatch[3]);
        if (buildDiff)
            return buildDiff;
        return thisMatch[4].localeCompare(configurationMatch[4]);
    }
    toParams() {
        let version_name = this.version_name;
        const ending = this.version_name ? this.version_name.substring(this.version_name.length - 2) : null;
        if (ending === ' E')
            version_name = this.version_name.substring(0, this.version_name.length - 2)
        return {
            platform: [this.platform],
            version:[this.version && !this.version_name ? Configuration.integerToVersion(this.version) : null],
            version_name: [version_name],
            is_simulator: [this.is_simulator === null ? null : (this.is_simulator ? 'True' : 'False')],
            style: [this.style],
            flavor: [this.flavor],
            model: [this.model],
            architecture: [this.architecture],
        };
    }
}

export {Configuration};

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/controller/configuration.py
# Copyright (C) 2019 Apple Inc. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1.  Redistributions of source code must retain the above copyright
#     notice, this list of conditions and the following disclaimer.
# 2.  Redistributions in binary form must reproduce the above copyright
#     notice, this list of conditions and the following disclaimer in the
#     documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS" AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import json

from resultsdbpy.flask_support.util import FlaskJSONEncoder


class Configuration(object):
    """
    This class is designed to use a partial configuration to match more complete configurations. Generally, an instance
    of this class will match another instance if all members match or are None. This means that, for example, a
    Configuration object with a platform of 'Mac' and all other members set to None will match any Configuration object
    with a platform of 'Mac', regardless of the values of the Configuration's other members.
    """

    VERSION_OFFSET_CONSTANT = 1000

    REQUIRED_MEMBERS = ['platform', 'is_simulator', 'version', 'architecture']
    OPTIONAL_MEMBERS = ['version_name', 'model', 'style', 'flavor']
    FILTERING_MEMBERS = ['sdk']

    @classmethod
    def from_json(cls, data):
        data = data if isinstance(data, dict) else json.loads(data)
        return Configuration(**{key: data.get(key) for key in cls.REQUIRED_MEMBERS + cls.OPTIONAL_MEMBERS + ['sdk']})

    def __init__(self, platform=None, version=None, sdk=None, version_name=None, is_simulator=None, architecture=None, model=None, style=None, flavor=None):
        self.platform = None if platform is None else str(platform)
        self.version = None if version is None else self.version_to_integer(version)
        self.version_name = None if version_name is None else str(version_name)
        self.is_simulator = None if is_simulator is None else bool(is_simulator)
        self.architecture = None if architecture is None else str(architecture)
        self.model = None if model is None else str(model)
        self.style = None if style is None else str(style)
        self.flavor = None if flavor is None else str(flavor)

        # Configurations which are identical, except for their sdk, should be considered identical
        self.sdk = None
        if sdk and sdk != '?':
            self.sdk = str(sdk)

    @classmethod
    def version_to_integer(cls, version):
        if isinstance(version, int):
            return version
        elif isinstance(version, str):
            version_list = version.split('.')
        elif isinstance(version, list) or isinstance(version, tuple):
            version_list = version
        else:
            raise TypeError(f'{type(version)} cannot be converted to a version integer')
        if len(version_list) < 1 or len(version_list) > 3:
            raise ValueError(f'{version} cannot be converted to a version integer')

        index = 0
        result = 0
        while index < 3:
            result *= cls.VERSION_OFFSET_CONSTANT
            if index < len(version_list):
                result += int(version_list[index])
            index += 1

        return result

    @classmethod
    def integer_to_version(cls, version):
        assert isinstance(version, int)
        result = [0, 0, 0]
        for index in range(3):
            result[-(index + 1)] = version % cls.VERSION_OFFSET_CONSTANT
            version //= cls.VERSION_OFFSET_CONSTANT
        return f'{result[0]}.{result[1]}.{result[2]}'

    def is_complete(self):
        return not any([getattr(self, member) is None for member in self.REQUIRED_MEMBERS])

    def __eq__(self, other):
        if not isinstance(other, type(self)):
            return False
        for member in self.REQUIRED_MEMBERS + self.OPTIONAL_MEMBERS:
            if getattr(self, member) is None or getattr(other, member) is None:
                continue
            if getattr(self, member) == getattr(other, member):
                continue
            return False
        return True

    def __ne__(self, other):
        return not (self == other)

    def __lt__(self, other):
        if not isinstance(other, type(self)):
            raise TypeError(f'Cannot compare {type(other)} and {type(self)}')
        for member in ['version', 'platform', 'version_name', 'is_simulator', 'model', 'architecture', 'style', 'flavor']:
            mine = getattr(self, member)
            theirs = getattr(other, member)
            if mine is None and theirs is None:
                continue
            if mine is None and theirs is not None:
                return True
            if mine is not None and theirs is None:
                return False
            if mine < theirs:
                return True
            if mine > theirs:
                return False
        return False

    def __le__(self, other):
        return self.__lt__(other) or self.__eq__(other)

    def __gt__(self, other):
        if not isinstance(other, type(self)):
            raise TypeError(f'Cannot compare {type(other)} and {type(self)}')
        for member in ['version', 'platform', 'version_name', 'is_simulator', 'model', 'architecture', 'style', 'flavor']:
            mine = getattr(self, member)
            theirs = getattr(other, member)
            if mine is None and theirs is None:
                continue
            if mine is None and theirs is not None:
                return False
            if mine is not None and theirs is None:
                return True
            if mine < theirs:
                return False
            if mine > theirs:
                return True
        return False

    def __ge__(self, other):
        return self.__gt__(other) or self.__eq__(other)

    def __hash__(self):
        result = 0
        for member in self.REQUIRED_MEMBERS + self.OPTIONAL_MEMBERS + self.FILTERING_MEMBERS:
            if member == 'version_name' and getattr(self, 'version'):
                continue
            result ^= hash(getattr(self, member))
        return result

    def __repr__(self):
        result = ''
        if self.platform is not None:
            result += ' ' + self.platform
        if self.version is not None:
            result += ' ' + self.integer_to_version(self.version)
        elif self.version_name is not None:
            result += ' ' + self.version_name
        if self.sdk:
            result += f' ({self.sdk})'
        if self.is_simulator is not None and self.is_simulator:
            result += ' Simulator'
        if self.style is not None:
            result += ' ' + self.style
        if self.flavor is not None:
            result += ' ' + self.flavor
        if self.model is not None:
            result += ' running on ' + self.model
        if self.architecture is not None:
            result += ' using ' + self.architecture

        return result[1:]

    def to_json(self, pretty_print=False):
        return json.dumps(self, cls=self.Encoder, sort_keys=pretty_print, indent=4 if pretty_print else None)

    def to_query(self):
        query = ''
        for member in self.REQUIRED_MEMBERS + self.OPTIONAL_MEMBERS + self.FILTERING_MEMBERS:
            value = getattr(self, member)
            if value is None:
                continue

            if member == 'version':
                query += f'{member}={self.integer_to_version(value)}&'
            elif member == 'is_simulator':
                query += f"{member}={'True' if value else 'False'}&"
            else:
                query += f'{member}={value}&'
        if query:
            return query[:-1]
        return ''

    class Encoder(FlaskJSONEncoder):

        def default(self, obj):
            if isinstance(obj, Configuration):
                result = {}
                for key, value in obj.__dict__.items():
                    if value is None:
                        continue
                    result[key] = value
                return result
            return super(Configuration.Encoder, self).default(obj)
